### ISSUE TITLE
2.12.1 - Scheduler undo keeps showing learning card until graduated

### ISSUE THREAD

**Issue by david-allison-1** [ankidroid/Anki-Android#6966]
SchevV1

This may have been a problem for some time, I recall seeing this before.

This does not match Anki Desktop 2.1.30

* Add a note with 3 cards
* review each as "Good"
* Undo "3"
* Answer "good"
* "3" is shown again

Anki Desktop 2.1.30 shows "1"

Patch is a test case based on 2.12.1

[Reviewer_-_undo_problem_patch.patch.txt](https://github.com/ankidroid/Anki-Android/files/5124981/Reviewer_-_undo_problem_patch.patch.txt)

**Comment by Arthur-Milchior** [comment]
«Patch format detection failed.» (both on both v2.12.1 and master)

Do you have a branch with it maybe, seems it'd be easier

**Comment by david-allison-1** [comment]
@Arthur-Milchior https://github.com/david-allison-1/Anki-Android/tree/2-12-1-undo

**Comment by Arthur-Milchior** [comment]
thanks

**Comment by Arthur-Milchior** [comment]
>java.lang.Exception: Main looper has queued unexecuted runnables. This might be the cause of the test failure. You might need a shadowOf(getMainLooper()).idle() call.

	at org.robolectric.android.internal.AndroidTestEnvironment.checkStateAfterTestFailure(AndroidTestEnvironment.java:470)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:548)
	at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$0(SandboxTestRunner.java:252)
	at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:89)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NoClassDefFoundError: org/bitbucket/cowwoc/diffmatchpatch/DiffMatchPatch
	at com.ichi2.utils.DiffEngine.<init>(DiffEngine.java:29)
	at com.ichi2.anki.AbstractFlashcardViewer.typeAnsAnswerFilter(AbstractFlashcardViewer.java:748)
	at com.ichi2.anki.AbstractFlashcardViewer.displayCardAnswer(AbstractFlashcardViewer.java:2041)
	at com.ichi2.anki.ReviewerTest.displayAnswer(ReviewerTest.java:208)
	at com.ichi2.anki.ReviewerTest.jsTime4ShouldBeBlankIfButtonUnavailable(ReviewerTest.java:79)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:546)
	... 6 more
Caused by: java.lang.ClassNotFoundException: org.bitbucket.cowwoc.diffmatchpatch.DiffMatchPatch
	at org.robolectric.internal.bytecode.SandboxClassLoader.getByteCode(SandboxClassLoader.java:164)
	at org.robolectric.internal.bytecode.SandboxClassLoader.maybeInstrumentClass(SandboxClassLoader.java:119)
	at org.robolectric.internal.bytecode.SandboxClassLoader.lambda$findClass$0(SandboxClassLoader.java:112)
	at org.robolectric.util.PerfStatsCollector.measure(PerfStatsCollector.java:53)
	at org.robolectric.internal.bytecode.SandboxClassLoader.findClass(SandboxClassLoader.java:111)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	... 23 more

For the second test. I'd like help.

For the first, I assume you unset bury siblings of new card, but I can't figure out where this occurs

**Comment by david-allison-1** [comment]
There's only one test:

`testMultipleCards`

```
java.lang.Exception: Main looper has queued unexecuted runnables. This might be the cause of the test failure. You might need a shadowOf(getMainLooper()).idle() call.

	at org.robolectric.android.internal.AndroidTestEnvironment.checkStateAfterTestFailure(AndroidTestEnvironment.java:470)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:548)
	at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$0(SandboxTestRunner.java:252)
	at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:89)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.AssertionError: Unexpected card ord
Expected: not is <3>
     but: was <3>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at com.ichi2.anki.ReviewerTest.assertCurrentOrdIsNot(ReviewerTest.java:134)
	at com.ichi2.anki.ReviewerTest.testMultipleCards(ReviewerTest.java:126)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:546)
	... 6 more


Process finished with exit code -1
```

**Comment by Arthur-Milchior** [comment]
`assertCounts(reviewer,0, 6, 0);` should probably fail if it's run less more 12 hours before the end of the day. Indeed, each card will have 2 or 3 cards to review today depending on the time remaining in the day.

**Comment by david-allison-1** [comment]
That's not what the test is failing with

**Comment by Arthur-Milchior** [comment]
I tried to apply it on 2.13, because it's complex to read 2.12, I don't have it in my mind anymore. And there, I did get a problem with count. If I don't check counts, the tests pass

**Comment by Arthur-Milchior** [comment]
My bad, the long error message was related to `jsTime4ShouldBeBlankIfButtonUnavailable` which is not new. I don't know why it's fail, but that's not the question of interest here

**Comment by Arthur-Milchior** [comment]
For information the error message I get is

> Expected: is "[0, 6, 0]"
>     but: was "[1, 4, 0]"
>	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
>	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:8)
>	at com.ichi2.anki.ReviewerTest.assertCounts(ReviewerTest.java:155)
>	at com.ichi2.anki.ReviewerTest.testMultipleCards(ReviewerTest.java:114)
>	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
>	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
>	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
>	at java.lang.reflect.Method.invoke(Method.java:498)
>	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
>	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
>	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
>	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
>	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
>	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
>	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
>	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:546)
>	... 6 more

This is why I mentioned the fact that the result of this test with change with the hour.

Indeed, when I get the last test to fail, I see a real problem that I'll try to understand

**Comment by Arthur-Milchior** [comment]
The bug was caused by d694006cb27bdecd13888f782e1248565c45e2bb . 
Essentially, the bug is due to the fact that I called `reset` while there is a card in the reviewer, without actually taking the card in the reviewer into account. Which is something I corrected later thanks to considering mHaveCounts and mHaveQueue separately.

**Comment by Arthur-Milchior** [comment]
This would explain https://github.com/ankidroid/Anki-Android/issues/6898 and https://github.com/ankidroid/Anki-Android/issues/6768

The idea here is that:
* the learn queue gets emptied during reset
* when the card is answered, it is added in the learn queue. But since the queue was empty and not yet refilled, it is added in a dirty queue. Normally, it's not a problem when the queue is dirty and empty, because `_fillLrn` will simply fill the queue again. However, as soon as there is an element in the queue, it is assumed that it is not dirty. This is corrected thanks to `mHaveQueue` which simply state whether the queue is dirty or not.
* Since the queue contains a single element and is assumed (incorrectly) not to be dirty, all other cards in learning never enter the queue, so they are reviewed at the end of the review session
* If there are only cards in learning, the single card in the queue is the only one being reviewed, until it goes to the Rev queue, at which point `fillLrn` is allowed to fill the queue again.


I believe that the more general problem here is that, while I did realize I corrected plenty of bugs on 2.1.13 alpha, I didn't realize that some of those bugs where in stable. Mostly because I have no idea when the stable branch was cut and that my mental model of the code uses the code in alpha.

**Comment by Arthur-Milchior** [comment]
the bug was corrected in 3ae67895c4667b6c83acef6b3b27945553354963 , but at this time I didn't even realize it was correcting a bug.

According to my git repo, this commit was created the 5th of July while the one introduced the bug was created the 30th of June. I'm not really surprised I don't have it in my mind anymore.

Regarding the creation of a 2.12.2 I can't tell which commit we would need to cherry pick exactly, because I don't know of which commit 3ae67895c4667b6c83acef6b3b27945553354963 depends. I'm pretty sure it can't be used alone , since it uses `mHaveCounts` which is not in 2.12.1

**Comment by david-allison-1** [comment]
Fixed by #6969
